### ISSUE TITLE
afterEach and beforeEach for property tests

### ISSUE THREAD

**Issue by hath995** [dubzzz/fast-check#169]
I'm trying to make a property based test based on actually manipulating the dom and then comparing two different dom states.

I'm using the mocha test runner. However, I'm having a problem with the state getting out of sync because the afterEach and beforeEach triggers for mocha do not trigger pre and post fast-check test as they're separate systems. So naturally I've moved most of the setup and tear down code into the property test predicate. This gets pretty cluttered. 

What do you think about adding hooks in the assert parameters for running setup/teardown? Something like the following. 
```
export interface Parameters<T = void> {
    seed?: number;      // optional, initial seed of the generator: Date.now() by default
    numRuns?: number;   // optional, number of runs before success: 100 by default
    maxSkipsPerRun?: number; // optional, maximal number of skipped entries per run: 100 by default
    timeout?: number;   // optional, only taken into account for asynchronous runs (asyncProperty)
                        // specify a timeout in milliseconds, maximum time for the predicate to return its result
                        // only works for async code, will not interrupt a synchronous code: disabled by default
    path?: string;      // optional, way to replay a failing property directly with the counterexample
                        // it can be fed with the counterexamplePath returned by the failing test (requires seed too)
    logger?: (v: string) => void; // optional, log output: console.log by default
    unbiased?: boolean; // optional, force the use of unbiased arbitraries: biased by default
    verbose?: boolean;  // optional, enable verbose mode: false by default
                        // when enabling verbose mode,
                        // you will be provided the list of all failing entries encountered
                        // whenever a property fails - useful to detect patterns
    examples?: T[];     // optional, custom values added to generated ones: [] by default
                        // when set, those examples will be run against the property first
                        // followed by the values generated by the framework
                        // they do not increase the number of times the property will be launched
    afterEach?: () => void; //runs teardown code after each property test
    beforeEach?: () => void; //runs setup code before each property test begins
}
```
Or maybe add them to the property

```
interface IProperty<Ts> {
    ...
    afterEach?: () => void; //runs teardown code after each property test
    beforeEach?: () => void; //runs setup code before each property test begins
}
```

It isn't a super important issue, but it might be a small quality of life feature.

**Comment by dubzzz** [comment]
Just drafted a quick and dirty implementation that should answer your needs.
You can see it in action at https://runkit.com/dubzzz/fastcheck-beforeeach-aftereach

```typescript
// Helper function to add support for beforeEach and afterEach

const wrapIt = ({test, beforeEach, afterEach}) => {
    return (...args) => {
        beforeEach(...args);
        
        let success = false;
        try { success = test(...args); }
        catch (err) {}
        
        afterEach(...args);
        return success;
    };
};

// How to use it?

fc.assert(
  fc.property(
    /* ...arbitraries */,
    wrapIt({
      beforeEach: /* beforeEach hook */,
      afterEach: /* afterEach hook */,
      test: /* test function checking that the property is OK */
    })
  )
);
```

Would it answer your needs?

I still need to see how to integrate it correctly into fast-check.
- Will it be an additional parameter in assert? Another kind of property? A withBeforeEach method on IProperty? A helper function?
- Does beforeEach/afterEach need to have access to the generated values?
- Do they need to be synchronous/asynchronous?
- etc...

**Comment by hath995** [comment]
Thanks, I did something similar to wrap it. It is sufficient for what I was doing. 

To answer the questions. 
1. I feel the most natural place is to make it a method of property. 
2. I don't think beforeEach need access to the value. If someone needed to do something with the value they should just use map/chain on an arbitrary or do it inside the property test. Having access to the value in afterEach might useful. I am reminded of when I went to a talk about phpUnit and how for some clients wanted to see logging for passing test case even for generated test cases. 
3. I would say that the beforeEach and after each should be asynchronous for `asyncProperty` and synchronous for `property`

I feel like it would feel similar to Promises then/catch/finally if it were a part of a property. 
```
fc.assert(
  fc.property(
    /* ...arbitraries */,
      test: /* test function checking that the property is OK */
    )
    .beforeEach(setupFn)
    .afterEach(tearDownFn)
  )
);
```

**Comment by dubzzz** [comment]
@hath995 I just started to implement a draft code for this issue. I am currently wondering what should be the input parameters of `setupFn` and `tearDownFn`. Do you have a clear idea for that?

I would be interested into having an example where there is such need for before and after before going ahead with the feature.

**Comment by dubzzz** [comment]
Regarding previous message, I just thought about one possible usage.

Let's say we want to test that the search field of google's homepage correctly preloads results on typing. We would basically define:

- before: go to www.google.com
- test: enter the text, check we see its associated results
- after: go to about:blank and close potential popups

**Comment by MehdiSaffar** [comment]
Is there any way to make every property test behave as if it is one seperate jest test? I have defined a few hooks like `beforeEach` to reset the database state and an `afterAll` to close the database connection. My property tests generate arbitrary data that gets inserted or updated in the database. I hit errors where the same primary key tries to get added, or where a row is updated but is not reverted. I have defined the hooks inside a `setupFilesAfterEnv`.
Basically is there a way to "revert" the nesting so that a jest test runs inside a property check, not the opposite? Thanks

**Comment by dubzzz** [comment]
@MehdiSaffar Whenever running a `fc.property`, `fc.assert` runs it multiple times and totally ignores the `beforeEach`/`afterEach` specified at test runner level (which seems to be the case in your case). In general fast-check is fully agnostic of the test runner in which it is executed so it cannot take into account the `beforeEach`/`afterEach` hooks defined at test runner level.

In other words, the following snippet will print:
```js
beforeEach(() => {
  console.log('beforeEach at test runner level');
});
afterEach(() => {
  console.log('afterEach at test runner level');
});
test('my test', () => {
  console.log('before fc.assert');
  fc.assert(
    fc.property(
      fc.nat(),
      fc.nat(),
      (a, b) => {
        console.log('property');
      }
    )
  )
  console.log('after fc.assert');
});

// It prints:
// > beforeEach at test runner level
// > before fc.assert
// > property (100x)
// > after fc.assert
// > afterEach at test runner level
```

If you want to run the `beforeEach`/`afterEach` before/after any call to the property, you'll need to update the code as follow:

```js
beforeEach(() => {
  console.log('beforeEach at test runner level');
});
afterEach(() => {
  console.log('afterEach at test runner level');
});
test('my test', () => {
  console.log('before fc.assert');
  fc.assert(
    fc.property(
      fc.nat(),
      fc.nat(),
      (a, b) => {
        console.log('property');
      }
    )
    .beforeEach(() => console.log('beforeEach at property level'))
    .afterEach(() => console.log('afterEach at property level'))
  )
  console.log('after fc.assert');
});

// It prints:
// > beforeEach at test runner level
// > before fc.assert
// > beforeEach at property level
// > property
// > afterEach at property level
// > beforeEach at property level
// > property
// > afterEach at property level
// > beforeEach at property level
// > property
// > afterEach at property level
// ...
// > after fc.assert
// > afterEach at test runner level
```

**Comment by MehdiSaffar** [comment]
@dubzzz I wasn't aware of the `beforeEach` and `afterEach` hooks of fast-check. Thanks a lot.

**Comment by CMCDragonkai** [comment]
If I'm using `testProp` from jest fast check plugin, is there an equivalent to hooking into the fast check's `beforeEach` and `afterEach`?

**Comment by dubzzz** [comment]
You can use: `fc.configureGlobal({beforeEach, afterEach})`. But otherwise so far, there is no such capability in `@fast-check/jest`. You should definitely open an issue so that it can be tracked
